gateway/api/ray.py

```diff
@@ -76,17 +77,48 @@ def submit(self, job: Job) -> Optional[str]:
         """
         tracer = trace.get_tracer("scheduler.tracer")
         with tracer.start_as_current_span("submit.job") as span:
+            # get program
             program = job.program
+
+            # get dependencies
             _, dependencies = try_json_loads(program.dependencies)
-            with tarfile.open(program.artifact.path) as file:
-                extract_folder = os.path.join(
-                    sanitize_file_path(str(settings.MEDIA_ROOT)),
-                    "tmp",
-                    str(uuid.uuid4()),
+
+            # get artifact
+            working_directory_for_upload = os.path.join(
+                sanitize_file_path(str(settings.MEDIA_ROOT)),
+                "tmp",
+                str(uuid.uuid4()),
+            )
+            if program.image is not None:
```

Comment:
Can you confirm me if I understood this logic, @IceKhan13 ? 🙏 

- We have a template as the entrypoint
- We fill the template with default data
- We upload the template in user's OS as the program

Reply:
yes, user implements `Runner` template and we are calling `Runner.run` from this `main.py` entrypoint.